Add Tableau workbook refresh via webhooks (#3123)
fishtown-analytics/docs.getdbt.com: website/docs/guides/orchestration/webhooks/zapier-refresh-tableau-workbook.md

---

> title: "Refresh a Tableau workbook with extracts when a job completes"

Reviewer: non blocker but sugg shortening title
```suggestion
title: "Refresh Tableau workbook with extracts after a job finishes"
```

id: webhooks-guide-zapier-refresh-tableau-workbook
slug: zapier-refresh-tableau-workbook
description: Use Zapier to trigger a Tableau workbook refresh
---

> This guide will teach you how to refresh a Tableau workbook that leverages extracts when a dbt Cloud job has completed successfully and there is fresh data available. The integration will:

Reviewer: sugg a link to tableau extracts
```suggestion
This guide will teach you how to refresh a Tableau workbook that leverages [extracts](https://help.tableau.com/current/pro/desktop/en-us/extracting_data.htm) when a dbt Cloud job has completed successfully and there is fresh data available. The integration will:
```

 - Receive a webhook notification in Zapier
 - Trigger a refresh of a Tableau workbook

## Prerequisites

> In order to set up the integration, you should have familiarity with:
> - [dbt Cloud Webhooks](/docs/deploy/webhooks)
> - Zapier
> - The [Tableau API](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api.htm)
> - The [version](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_concepts_versions.htm#rest_api_versioning) of Tableau's REST API that is compatible with your server 

Reviewer: ```suggestion
To set up the integration, you need to be familiar with:

- [dbt Cloud Webhooks](/docs/deploy/webhooks)
- Zapier
- The [Tableau API](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api.htm)
- The [version](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_concepts_versions.htm#rest_api_versioning) of Tableau's REST API that is compatible with your server 
```

## Integration steps

### 1. Obtain authentication credentials from Tableau

> Obtain a [Personal Access Token](https://help.tableau.com/current/server/en-us/security_personal_access_tokens.htm) from your Tableau Server/Cloud instance.  You'll need this to authenticate with the Tableau API.  In addition, you'll want to ensure that your Tableau workbook uses data sources that allow refresh access.  This setting is typically set when publishing.

Reviewer: ```suggestion
To authenticate with the Tableau API, obtain a [Personal Access Token](https://help.tableau.com/current/server/en-us/security_personal_access_tokens.htm) from your Tableau Server/Cloud instance. In addition, make sure your Tableau workbook uses data sources that allow refresh access, which is usually set when publishing.
```

### 2. Create a new Zap in Zapier

> Use **Webhooks by Zapier** as the Trigger, and **Catch Raw Hook** as the Event. If you don't intend to [validate the authenticity of your webhook](docs/deploy/webhooks#validate-a-webhook) (not recommended!) then you can choose **Catch Hook** instead. 
> 
> Press **Continue**, then copy the webhook URL. 

Reviewer: sugg clarifying intro to what user is doing here

```suggestion

To set up a webhook in Zapier, select **Webhooks by Zapier** as the Trigger and **Catch Raw Hook** as the Event. However, if you choose not to [validate the authenticity of your webhook](docs/deploy/webhooks#validate-a-webhook), which isn't recommended, you can choose **Catch Hook** instead. 

Press **Continue** and then copy the webhook URL. 
```

![Screenshot of the Zapier UI, showing the webhook URL ready to be copied](/img/guides/orchestration/webhooks/zapier-common/catch-raw-hook.png)

### 3. Configure a new webhook in dbt Cloud
See [Create a webhook subscription](/docs/deploy/webhooks#create-a-webhook-subscription) for full instructions. Your event should be **Run completed**, and you need to change the **Jobs** list to only contain any jobs whose completion should trigger a report refresh.

Make note of the Webhook Secret Key for later.  Paste in the webhook URL obtained from Zapier in step 2 into the "Endpoint" field and test the endpoint.

Once you've tested the endpoint in dbt Cloud, go back to Zapier and click **Test Trigger**, which will create a sample webhook body based on the test event dbt Cloud sent.

> The sample body's values are hard-coded and not reflective of your project, but they give Zapier a correctly-shaped object during development. 

Reviewer: ```suggestion
To set up a webhook subscription for dbt Cloud, follow the instructions in [Create a webhook subscription](/docs/deploy/webhooks#create-a-webhook-subscription). For the event, choose **Run completed** and modify the **Jobs** list to include only the jobs that should trigger a report refresh.

Remember to save the Webhook Secret Key for later. Paste in the webhook URL obtained from Zapier in step 2 into the **Endpoint** field and test the endpoint.

Once you've tested the endpoint in dbt Cloud, go back to Zapier and click **Test Trigger**, which will create a sample webhook body based on the test event dbt Cloud sent.

The sample body's values are hard-coded and not reflective of your project, but they give Zapier a correctly-shaped object during development. 
```

Zapier allows you to [store secrets](https://help.zapier.com/hc/en-us/articles/8496293271053-Save-and-retrieve-data-from-Zaps), which prevents your keys from being displayed in plaintext in the Zap code. You will be able to access them via the [StoreClient utility](https://help.zapier.com/hc/en-us/articles/8496293969549-Store-data-from-code-steps-with-StoreClient).

This guide assumes the names for the secret keys are: `DBT_WEBHOOK_KEY`, `TABLEAU_SITE_URL`, `TABLEAU_SITE_NAME`, `TABLEAU_API_TOKEN_NAME`, and `TABLEAU_API_TOKEN_SECRET`. If you are using different names, make sure you update all references to them in the sample code.

This guide uses a short-lived code action to store the secrets, but you can also use a tool like Postman to interact with the [REST API](https://store.zapier.com/) or create a separate Zap and call the [Set Value Action](https://help.zapier.com/hc/en-us/articles/8496293271053-Save-and-retrieve-data-from-Zaps#3-set-a-value-in-your-store-0-3).

> #### a. Create a Storage by Zapier connection
> If you haven't already got one, go to <https://zapier.com/app/connections/storage> and create a new connection. Remember the UUID secret you generate for later. 

Reviewer: ```suggestion
#### a. Create a Storage by Zapier connection 

Create a new connection at https://zapier.com/app/connections/storage if you don't already have one and remember the UUID secret you generate for later.
```

> #### b. Add a temporary code step
> Choose **Run Python** as the Event. Run the following code: 
> ```python 
> store = StoreClient('abc123') #replace with your UUID secret
> store.set('DBT_WEBHOOK_KEY', 'abc123') #replace with your dbt Cloud Webhook key
> store.set('TABLEAU_SITE_URL', 'abc123') #replace with your Tableau Site URL, inclusive of https:// and .com
> store.set('TABLEAU_SITE_NAME', 'abc123') #replace with your Tableau Site/Server Name
> store.set('TABLEAU_API_TOKEN_NAME', 'abc123') #replace with your Tableau API Token Name
> store.set('TABLEAU_API_TOKEN_SECRET', 'abc123') #replace with your Tableau API Secret
> ```
> Test the step. You can delete this Action when the test succeeds. The key will remain stored as long as it is accessed at least once every three months.

Reviewer: ```suggestion
#### b. Add a temporary code step

Choose **Run Python** as the Event and run the following code: 

```python 
store = StoreClient('abc123') #replace with your UUID secret
store.set('DBT_WEBHOOK_KEY', 'abc123') #replace with your dbt Cloud Webhook key
store.set('TABLEAU_SITE_URL', 'abc123') #replace with your Tableau Site URL, inclusive of https:// and .com
store.set('TABLEAU_SITE_NAME', 'abc123') #replace with your Tableau Site/Server Name
store.set('TABLEAU_API_TOKEN_NAME', 'abc123') #replace with your Tableau API Token Name
store.set('TABLEAU_API_TOKEN_SECRET', 'abc123') #replace with your Tableau API Secret
```
Test the step. You can delete this Action when the test succeeds. The key will remain stored as long as it is accessed at least once every three months.
```

> ### 4. Add a code action

Reviewer: duplicate 4 so changing this to 5

```suggestion
### 5. Add a code action
```

Select **Code by Zapier** as the App, and **Run Python** as the Event. 

In the **Set up action** area, add two items to **Input Data**: `raw_body` and `auth_header`. Map those to the `1. Raw Body` and `1. Headers Http Authorization` fields from the **Catch Raw Hook** step above.

![Screenshot of the Zapier UI, showing the mappings of raw_body and auth_header](/img/guides/orchestration/webhooks/zapier-common/run-python.png)

In the **Code** field, paste the following code, replacing `YOUR_STORAGE_SECRET_HERE` in the StoreClient constructor with the UUID secret you created when setting up the Storage by Zapier integration, and replacing the `workbook_name` and `api_version` variables to actual values.

> The code below will validate the authenticity of the request then obtain the workbook ID for the specified workbook name.  Next, the code will send a [`update workbook` command to the Tableau API](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_ref_workbooks_and_views.htm#update_workbook_now) for the given workbook ID.

Reviewer: ```suggestion
The following code validates the authenticity of the request and obtains the workbook ID for the specified workbook name.  Next, the code will send a [`update workbook` command to the Tableau API](https://help.tableau.com/current/api/rest_api/en-us/REST/rest_api_ref_workbooks_and_views.htm#update_workbook_now) for the given workbook ID.
```

#Extract token to use for subsequent calls
auth_token = auth_response.json()["credentials"]["token"]
site_id = auth_response.json()["credentials"]["site"]["id"]

#Extract the workbook ID
workbooks_url = f"{server_url}/api/{api_version}/sites/{site_id}/workbooks"
workbooks_headers = {
    "Accept": "application/json",
    "Content-Type": "application/json",
    "X-Tableau-Auth": auth_token
}
workbooks_params = {
    "filter": f"name:eq:{workbook_name}"
}
workbooks_response = requests.get(workbooks_url, headers=workbooks_headers, params=workbooks_params)

#Assign workbook ID
workbooks_data = workbooks_response.json()
workbook_id = workbooks_data["workbooks"]["workbook"][0]["id"]

# Refresh the workbook
refresh_url = f"{server_url}/api/{api_version}/sites/{site_id}/workbooks/{workbook_id}/refresh"
refresh_data = {}
refresh_headers = {
    "Accept": "application/json",
    "Content-Type": "application/json",
    "X-Tableau-Auth": auth_token
}

refresh_trigger = requests.post(refresh_url, data=json.dumps(refresh_data), headers=refresh_headers)
return {"message": "Workbook refresh has been queued"}
```

> ### 5. Test and deploy

Reviewer: Update step
```suggestion
### 6. Test and deploy
```

> You can iterate on the Code step by modifying the code and then running the test again. When you're happy with it, you can publish your Zap.

Reviewer: ```suggestion
To make changes to your code, you can modify it and test it again. When you're happy with it, you can publish your Zap.
```

This guide uses a short-lived code action to store the secrets, but you can also use a tool like Postman to interact with the [REST API](https://store.zapier.com/) or create a separate Zap and call the [Set Value Action](https://help.zapier.com/hc/en-us/articles/8496293271053-Save-and-retrieve-data-from-Zaps#3-set-a-value-in-your-store-0-3).

#### a. Create a Storage by Zapier connection 

Create a new connection at https://zapier.com/app/connections/storage if you don't already have one and remember the UUID secret you generate for later.

#### b. Add a temporary code step

Choose **Run Python** as the Event and input the following code: 

```python 
store = StoreClient('abc123') #replace with your UUID secret
store.set('DBT_WEBHOOK_KEY', 'abc123') #replace with your dbt Cloud Webhook key
store.set('TABLEAU_SITE_URL', 'abc123') #replace with your Tableau Site URL, inclusive of https:// and .com
store.set('TABLEAU_SITE_NAME', 'abc123') #replace with your Tableau Site/Server Name
store.set('TABLEAU_API_TOKEN_NAME', 'abc123') #replace with your Tableau API Token Name
store.set('TABLEAU_API_TOKEN_SECRET', 'abc123') #replace with your Tableau API Secret

> 

Reviewer 2: Missing closing triple backticks here

Author: Thanks @joellabes .  Resolved.
